Index: compat-wireless-2013-04-16/drivers/net/wireless/rt2x00/rt2800lib.c
===================================================================
--- compat-wireless-2013-04-16.orig/drivers/net/wireless/rt2x00/rt2800lib.c	2013-06-01 11:51:08.000000000 +0400
+++ compat-wireless-2013-04-16/drivers/net/wireless/rt2x00/rt2800lib.c	2013-06-01 11:55:56.000000000 +0400
@@ -3045,6 +3045,11 @@
 	} else {
 		power_ctrl = 0;
 	}
+
+	if (rt2x00_rt(rt2x00dev, RT5350)) {
+		power_ctrl = 3;
+	}
+
 	rt2x00_set_field8(&r1, BBP1_TX_POWER_CTRL, power_ctrl);
 	rt2800_bbp_write(rt2x00dev, 1, r1);
 	offset = TX_PWR_CFG_0;
@@ -3537,13 +3542,14 @@
 	} else if (rt2x00_rt(rt2x00dev, RT3572)) {
 		rt2800_register_write(rt2x00dev, TX_SW_CFG0, 0x00000400);
 		rt2800_register_write(rt2x00dev, TX_SW_CFG1, 0x00080606);
-	} else if (rt2x00_rt(rt2x00dev, RT5350) ||
-		   rt2x00_rt(rt2x00dev, RT5390) ||
+	} else if (rt2x00_rt(rt2x00dev, RT5390) ||
 		   rt2x00_rt(rt2x00dev, RT5392) ||
 		   rt2x00_rt(rt2x00dev, RT5592)) {
 		rt2800_register_write(rt2x00dev, TX_SW_CFG0, 0x00000404);
 		rt2800_register_write(rt2x00dev, TX_SW_CFG1, 0x00080606);
 		rt2800_register_write(rt2x00dev, TX_SW_CFG2, 0x00000000);
+	} else if (rt2x00_rt(rt2x00dev, RT5350)) {
+		rt2800_register_write(rt2x00dev, TX_SW_CFG0, 0x00000404);
 	} else {
 		rt2800_register_write(rt2x00dev, TX_SW_CFG0, 0x00000000);
 		rt2800_register_write(rt2x00dev, TX_SW_CFG1, 0x00080606);
@@ -4316,6 +4322,7 @@
 		rt2800_bbp_write(rt2x00dev, 143, 0xa2);
 
 		rt2800_bbp_write(rt2x00dev, 148, 0xc8);
+		rt2800_bbp_write(rt2x00dev, 70, 0xa);
 	}
 
 	if (rt2x00_rt(rt2x00dev, RT5350)) {
