Index: compat-wireless-2013-01-07/drivers/net/wireless/rt2x00/rt2800lib.c
===================================================================
--- compat-wireless-2013-01-07.orig/drivers/net/wireless/rt2x00/rt2800lib.c	2013-02-15 18:58:52.297563645 +0400
+++ compat-wireless-2013-01-07/drivers/net/wireless/rt2x00/rt2800lib.c	2013-02-15 19:34:49.136356059 +0400
@@ -2661,6 +2661,11 @@
 	} else {
 		power_ctrl = 0;
 	}
+
+	if (rt2x00_rt(rt2x00dev, RT5350)) {
+		power_ctrl = 3;
+	}
+
 	rt2x00_set_field8(&r1, BBP1_TX_POWER_CTRL, power_ctrl);
 	rt2800_bbp_write(rt2x00dev, 1, r1);
 	offset = TX_PWR_CFG_0;
@@ -3138,12 +3138,13 @@
 	} else if (rt2x00_rt(rt2x00dev, RT3572)) {
 		rt2800_register_write(rt2x00dev, TX_SW_CFG0, 0x00000400);
 		rt2800_register_write(rt2x00dev, TX_SW_CFG1, 0x00080606);
-	} else if (rt2x00_rt(rt2x00dev, RT5350) ||
-		   rt2x00_rt(rt2x00dev, RT5390) ||
+	} else if (rt2x00_rt(rt2x00dev, RT5390) ||
 		   rt2x00_rt(rt2x00dev, RT5392)) {
 		rt2800_register_write(rt2x00dev, TX_SW_CFG0, 0x00000404);
 		rt2800_register_write(rt2x00dev, TX_SW_CFG1, 0x00080606);
 		rt2800_register_write(rt2x00dev, TX_SW_CFG2, 0x00000000);
+	} else if (rt2x00_rt(rt2x00dev, RT5350)) {
+		rt2800_register_write(rt2x00dev, TX_SW_CFG0, 0x00000404);
 	} else {
 		rt2800_register_write(rt2x00dev, TX_SW_CFG0, 0x00000000);
 		rt2800_register_write(rt2x00dev, TX_SW_CFG1, 0x00080606);
@@ -3725,7 +3726,6 @@
 	    rt2x00_rt(rt2x00dev, RT3090) ||
 	    rt2x00_rt(rt2x00dev, RT3390) ||
 	    rt2x00_rt(rt2x00dev, RT3572) ||
-	    rt2x00_rt(rt2x00dev, RT5350) ||
 	    rt2x00_rt(rt2x00dev, RT5390) ||
 	    rt2x00_rt(rt2x00dev, RT5392)) {
 		rt2800_bbp_read(rt2x00dev, 138, &value);
@@ -3783,6 +3783,7 @@
 		rt2800_bbp_write(rt2x00dev, 143, 0xa2);
 
 		rt2800_bbp_write(rt2x00dev, 148, 0xc8);
+		rt2800_bbp_write(rt2x00dev, 70, 0xa);
 	}
 
 	if (rt2x00_rt(rt2x00dev, RT5350)) {
